PurchasesCoreSwift/Networking/Backend.swift

```diff
@@ -0,0 +1,614 @@
+            var extraUserInfo = [ErrorDetails.finishableKey: NSNumber(value: finishable)] as [String: AnyObject]
+            extraUserInfo.merge(subscriberAttributesErrorInfo) { _, new in new }
+            let code = maybeNumberFromError(code: response?["code"])
+            let message = response?["message"] as? String
+            let responseError = ErrorUtils.backendError(withBackendCode: code,
+                                                        backendMessage: message,
+                                                        extraUserInfo: extraUserInfo as [NSError.UserInfoKey: Any])
+            completion(maybePurchaserInfo, responseError)
+            return
+        } else {
+            completion(maybePurchaserInfo, nil)
+            return
+        }
+    }
+
+    private func escapedAppUserID(appUserID: String) -> String {
+        return appUserID.addingPercentEncoding(withAllowedCharacters: NSMutableCharacterSet.urlHostAllowed)!
+    }
+
+    private func subscriberAttributesToDict(subscriberAttributes: SubscriberAttributeDict) -> [String: AnyObject] {
+        var attributesByKey: [String: AnyObject] = [:]
+        for (key, value) in subscriberAttributes {
+            attributesByKey[key] = value.asBackendDictionary() as AnyObject
+        }
+        return attributesByKey
+    }
+
+    private func userInfoAttributes(response: [String: AnyObject], statusCode: Int) -> [String: AnyObject] {
+        var resultDict: [String: AnyObject] = [:]
+
+        let isInternalServerError = statusCode >= HTTPStatusCodes.internalServerError.rawValue
+        let isNotFoundError = statusCode == HTTPStatusCodes.notFoundError.rawValue
+        let successfullySynced = !(isInternalServerError || isNotFoundError)
+        resultDict[Backend.RCSuccessfullySyncedKey as String] = NSNumber(value: successfullySynced)
+
+        let hasAttributesResponseContainerKey = (response[Backend.RCAttributeErrorsResponseKey] != nil)
+        let maybeAttributesResponseDict = hasAttributesResponseContainerKey
+            ? response[Backend.RCAttributeErrorsResponseKey] as? [String: AnyObject]
+            : response
+
+        if let attributesResponseDict = maybeAttributesResponseDict {
+            resultDict[Backend.RCAttributeErrorsKey] = attributesResponseDict[Backend.RCAttributeErrorsKey]
+        }
+
+        return resultDict
+    }
+
+    // MARK: Callback cache management
+
+    private func add(callback: @escaping BackendPurchaserInfoResponseHandler, key: String) -> Bool {
+        return callbackQueue.sync { [self] in
+            let maybeCallbacks = purchaserInfoCallbacksCache[key]
+            var callbacks: [BackendPurchaserInfoResponseHandler]
+            let requestAlreadyInFlight: Bool
+            if maybeCallbacks == nil {
+                requestAlreadyInFlight = false
+                callbacks = []
+            } else {
+                requestAlreadyInFlight = true
+                callbacks = maybeCallbacks!
+            }
```

Comment:
[Just a suggestion] In order to avoid the force unwrapping (maybe in the future, the rule `force_unwrapping` could be activated), we could use `if let` instead. WDYT?
```suggestion
            if let someCallbacks = maybeCallbacks {
                requestAlreadyInFlight = true
                callbacks = callbacks
            } else {
                requestAlreadyInFlight = false
                callbacks = []
            }
```

Reply:
I agree, in general `force_unwrapping` is something to be avoided. However, if you can prove that for unwrapping won't cause problems, it's ok. That being said, in this case, your suggestion removes the need for force unwrapping and is better, so I'll go with that!

Reply:
totally agree. Feel free to leave this code there 🙂